src/backend/cdb/cdbmutate.c

```diff
@@ -438,6 +445,13 @@ apply_motion(PlannerInfo *root, Plan *plan, Query *query)
 			/* If the query comes from 'CREATE TABLE AS' or 'SELECT INTO' */
 			if (query->parentStmtType != PARENTSTMTTYPE_NONE)
 			{
+			    contain_node_context    mod_ctx;
+
+			    planner_init_plan_tree_base(&mod_ctx.base, root);
+			    if (contain_modifyTable_walker((Node *)plan, &mod_ctx))
```

Comment:
Maybe simply check `query->hasModifyingCTE`? It would be better because ORCA-plan will be covered too (I suspect that ORCA can't plan such queries now). ORCA uses DML node instead of Modify Table node.

Reply:
What about a case when CTAS and Modify Table are executed in the single writer slice? Is it possible?

Reply:
I thought ORCA creates `PlannedStmt` independently from Postgres  Optimizer, where the cdbparallelize() is called. However, I share the `query->hasModifyingCTE` approach, because it is the only case with such queries.

Reply:
I found the way to hold 2 modifications in one slice. However, some changes are required.
First of all, we can set the target distribution policy of newly created table:
```
create table with_dml_dr (i int, j int) distributed replicated;
create table t_new as (with cte as
(delete from with_dml_dr returning *)
select i from cte) distributed replicated;
```
In current implementation the result of CTE has FLOW_REPLICATED and the same locus type. And in `apply_motion` function we have the [followiing](https://github.com/arenadata/gpdb/blob/2de3b73a03045062849da9b99612bcd022334a7b/src/backend/cdb/cdbmutate.c#L560-L579) picture: TargetPolicy is Replicated, and the desicion for Broadcast is made. Currently, FLOW_REPLICATED is not supported, therefore, the `needToBroadcast` is true` and motion is added:
```
explain (costs off)
create table t_new as (with cte as
(delete from with_dml_dr returning *)
select i from cte) distributed replicated;
                 QUERY PLAN                  
---------------------------------------------
 Broadcast Motion 3:3  (slice1; segments: 3)
   ->  Subquery Scan on cte
         ->  Delete on with_dml_dr
               ->  Seq Scan on with_dml_dr
 Optimizer: Postgres query optimizer
(5 rows)
create table t_new as (with cte as
(delete from with_dml_dr returning *)
select i from cte) distributed replicated;
ERROR:  INSERT/UPDATE/DELETE must be executed by a writer segworker group (nodeModifyTable.c:1761)  (seg2 slice1 127.0.0.1:6004 pid=570164) (nodeModifyTable.c:1761)
```
But, if we support FLOW_REPLICATED in this case, for example, let's set `needToBroadcast` to `false`:
```
insert into with_dml_dr values (1,1), (2,2), (3,3);
INSERT 0 3
explain (costs off) 
create table t_new as (with cte as
(delete from with_dml_dr returning *)
select i from cte) distributed replicated;
             QUERY PLAN              
-------------------------------------
 Subquery Scan on cte
   ->  Delete on with_dml_dr
         ->  Seq Scan on with_dml_dr
 Optimizer: Postgres query optimizer
(4 rows)
create table t_new as (with cte as
(delete from with_dml_dr returning *)
select i from cte) distributed replicated;
SELECT 1
select * from with_dml_dr;
 i | j 
---+---
(0 rows)
select * from t_new;
 i 
---
 1
 2
 3
(3 rows)
```

Reply:
So, as for hashed tables, and other operations the addition of Motion node is guaranteed, because the result of RETURNING for any DML operation [has](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/createplan.c#L6907) Strewn distribution. Therefore, in order to create the hashed table in CTAS, the redistribution is required.
As for replicated tables, the redistribution will be failed, because the `repartitionPlan` function will fail with the [assertion](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/cdb/cdbllize.c#L1185-L1186) error.

Another example when target distribution is Replicated (and result of CTE is Strewn):
```
explain (costs off)
create table t_new as (with cte as
(delete from with_dml returning *)
select i from cte) distributed replicated;
                 QUERY PLAN                  
---------------------------------------------
 Broadcast Motion 3:3  (slice1; segments: 3)
   ->  Subquery Scan on cte
         ->  Delete on with_dml
               ->  Seq Scan on with_dml
 Optimizer: Postgres query optimizer
(5 rows)

```

Reply:
After the voice chat, we decide to stop on a simple solution with `query->hasModifyingCTE`, which can be modified in future if needed.